### PR TITLE
[DISCUSS] Add small overview of modern C++ features utilized in Serac + references

### LINKED ISSUE #246: Add modern C++ references to the docs
We should add links to good modern C++ references in our docs for code developers who are not used to the features we use heavily in Serac (smart pointers, `std::variant`, `std::optional`, etc).

### PR DESCRIPTION
**Pull request by joshessman-llnl**
I wonder if it would also be useful to provide examples of the mentioned features being used in the code and explain the motivation behind their use.

The Meyers book I mention at the bottom is a great resource for the C++11/14 stuff but the articles I've linked are fairly good as well.

Resolves #246

### CODE REVIEW

File: src/docs/sphinx/developer_guide.rst [hunk 1]
```diff
@@ -61,6 +61,41 @@ Important Data Structures
...
+``std::unique_ptr<T>`` is used to denote **exclusive** ownership of a pointer to ``T`` - see `this article <https://www.drdobbs.com/cpp/c11-uniqueptr/240002708>`_ for more info.
+Because ``unique_ptr`` implies unique/exclusive ownership, instances cannot be copied.  For example, if a function has a ``unique_ptr`` argument, a caller must utilize
+*move semantics* to transfer ownership at the call site.  The linked article provides an example of this, and move semantics are discussed in a more general sense 
+`here <https://herbsutter.com/2020/02/17/move-simply/>`_.
+
+``std::shared_ptr<T>`` is used to denote **shared** ownership of a pointer to ``T`` - see `this article <https://docs.microsoft.com/en-us/cpp/cpp/how-to-create-and-use-shared-ptr-instances?view=vs-2019>`_ for example uses.
+``shared_ptr`` s should be used sparingly.  Often, when two objects need to share a resource, it is sufficient for only one of the objects to 
+be responsible for the lifetime of the shared resource; the other object can store a reference to the resource.
+
+``std::optional<T>`` is used to express the idea of ``Maybe T``, a.k.a. a nullable type.  An ``optional`` is optionally a ``T``,
```
Comment (jamiebramwell) [inline src/docs/sphinx/developer_guide.rst:79]: I would mention this frequently replaces checking for `nullptr` in legacy C++.

**Review by jamiebramwell — APPROVED** [review]
Thanks for typing this up! I think it will help a bunch.

**Comment by jamiebramwell** [comment]
I think the [Microsoft lambda overview](https://docs.microsoft.com/en-us/cpp/cpp/lambda-expressions-in-cpp) is pretty good.

**Comment by jamiebramwell** [comment]
@samuelpmishLLNL , can you take a look at this?

**Review by samuelpmishLLNL — APPROVED** [review]
looks good to me!

**Comment by joshessman-llnl** [comment]
LGTM

**Pull request merged**